### PR TITLE
core[minor]: Add maxsize for InMemoryCache

### PR DESCRIPTION
**Pull request by spyrosavl**
This PR introduces a maxsize parameter for the InMemoryCache class, allowing users to specify the maximum number of items to store in the cache. If the cache exceeds the specified maximum size, the oldest items are removed. Additionally, comprehensive unit tests have been added to ensure all functionalities are thoroughly tested. The tests are written using pytest and cover both synchronous and asynchronous methods.

Twitter: @spyrosavl

### CODE REVIEW

**Comment by spyrosavl** [comment]
I can't find exactly where are the docs to edit. If I understand correctly the doc for InMemoryCache [is still under the community package](https://api.python.langchain.com/en/latest/cache/langchain_community.cache.InMemoryCache.html#langchain_community.cache.InMemoryCache). That shouldn't be the case, right? Can you please help with this?

PS it's my first PR :)

File: libs/core/langchain_core/caches.py [hunk 1]
```diff
@@ -144,16 +145,27 @@ async def aclear(self, **kwargs: Any) -> None:
...
+            raise ValueError("maxsize must be greater than 0")
+        self._maxsize = maxsize
 
     def lookup(self, prompt: str, llm_string: str) -> Optional[RETURN_VAL_TYPE]:
         """Look up based on prompt and llm_string."""
         return self._cache.get((prompt, llm_string), None)
 
     def update(self, prompt: str, llm_string: str, return_val: RETURN_VAL_TYPE) -> None:
         """Update cache based on prompt and llm_string."""
+        if self._maxsize is not None and len(self._cache) >= self._maxsize:
```
Comment (nfcampos) [inline libs/core/langchain_core/caches.py:167]: should this be `while`?

Comment (spyrosavl) [reply libs/core/langchain_core/caches.py:167]: Why `while`? One item is inserted and one is deleted, no?

Comment (nfcampos) [reply libs/core/langchain_core/caches.py:167]: Then the check should be `==` not `>=` ?

Comment (spyrosavl) [reply libs/core/langchain_core/caches.py]: Fixed!

File: libs/core/langchain_core/caches.py [hunk 2]
```diff
@@ -144,16 +145,27 @@ async def aclear(self, **kwargs: Any) -> None:
 class InMemoryCache(BaseCache):
     """Cache that stores things in memory."""
 
-    def __init__(self) -> None:
-        """Initialize with empty cache."""
+    def __init__(self, maxsize: Optional[int] = None) -> None:
```
Comment (eyurtsev) [inline libs/core/langchain_core/caches.py:148]:
```suggestion
    def __init__(self, *, maxsize: Optional[int] = None) -> None:
```

**Review by eyurtsev — APPROVED** [review]

**Comment by eyurtsev** [comment]
@spyrosavl i'll resolve conflicts right now

**Comment by spyrosavl** [comment]
@eyurtsev I saw the linter failed for 3.8 cause I used `tuple` instead of `Tuple` in the test. Fixed it.

File: libs/core/tests/unit_tests/caches/test_in_memory_cache.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+def test_clear(cache: InMemoryCache) -> None:
+    """Test the clear method of InMemoryCache."""
+    prompt, llm_string, generations = cache_item(1)
+    cache.update(prompt, llm_string, generations)
+    cache.clear()
+    assert cache.lookup(prompt, llm_string) is None
+
+
+@pytest.mark.asyncio
```
Comment (eyurtsev) [inline libs/core/tests/unit_tests/caches/test_in_memory_cache.py:82]: We use auto-mode so, you don't need to type @pytest.mark.asyncio :)

Comment (spyrosavl) [reply libs/core/tests/unit_tests/caches/test_in_memory_cache.py]: Oh nice, I removed it!

**Comment by spyrosavl** [comment]
@eyurtsev do you have any clue why the linter is failing?

**Comment by eyurtsev** [comment]
I'll check when back at a computer

**Pull request merged**